portfolio-project-matching-app/pages/newProject.js

```diff
@@ -1,40 +1,114 @@
+            });
+            // set state if component is still mounted
+            if (isMounted) setAllTechnologies(technologies);
+        })
+        // cleanup function to assign false to isMounted
+        return function cleanup() {
+            isMounted = false
+        }
+    }, [])
+
     const addProject = async (e) => {
+        /*
+        DESCRIPTION:    upon submitting the new project form, this function
+                        will add the new project (and associations) to Firebase
+
+        INPUT:          e (event): passed implicitly by form onSubmit
+
+        RETURN:         NA
+        */
+        // prevent page from reloading
         e.preventDefault();
-        console.log(e.target.technologies);
-        let payload = {
-            name: e.target.name.value,
-            description: e.target.description.value,
-            capacity: e.target.capacity.value,
-            technologies: selectedTechnologies,
-            cencus: 1,
-            open: true,
-            likes: 0
+        // test payload for missing form data
+        if (payload.name === '' || payload.description === '' || payload.capacity === '') {
+            alert('Please fill out entire form');
+        }
+        // handle case where user is not logged in
+        else if (!authUser.user) {
+            alert('Please login to create a project');
+        }
+        // handle case of valid input
+        else {
+            // populate payload with data not captured in form
+            buildAndSetPayload();
+            // create project document in Firebase
+            const projectSnap = await createDoc('projects', payload);
+            // create project/technology associations
+            for (const technology of selectedTechnologies) {
+                await createAssociation('projects_technologies', projectSnap.id, technology.id);
+            }
+            // create user/owner association
+            await createAssociation('projects_users', projectSnap.id, authUser.user.id);
+            // flush state
+            updateSelectedTechnologies([]);
+            setPayload({name: '', description: '', capacity: '', });
         }
-        console.log(payload)
-        // await addNewDoc('projects', payload);
-        updateSelectedTechnologies([]);
+    }
+
+    const buildAndSetPayload = () => {
+        /*
+        DESCRIPTION:    Populates payload fields not called for explicitly by
+                        the form e.g., ownerId, census, etc.
+
+        INPUT:          NA
+
+        RETURN:         NA
+        */
+        // get open field value by examining capacity
+        const openStatus = payload.capacity === 1 ? false : true;
+        // build on payload object with values from form
+        // there has to be a better way to do this
+        const newPayload = payload;
+        newPayload.census = 1;
+        newPayload.ownerId = authUser.user.id;
+        newPayload.open = openStatus
```

Comment:
There **HAS** to be a better way to do this. I've tried the shorthand
`setPayload({...newPayload, ...{new keys and values}})`
but it doesn't seem to successfully set the state. Actually, I've tried a few variations of the above code snippet, but to no avail.

Reply:
I think the following would work:
```JavaScript
setPayload({
  ...payload,
  census: 1,
  ownerId: authUser.user.id,
  open: openStatus
})
```

Reply:
But it's the same number of lines so doesn't really matter.

Reply:
I swear I tried that, but for whatever reason the `payload` remained unchanged. It was almost behaving as if `setpayload` is async and I wasn't awaiting it.

Reply:
I even tested it out in my `test_dao` and it would work with dummy objects being combined, but then I couldn't get it to work in practice. I was pulling out my remaining hair.